api4/websocket.go

```diff
@@ -36,53 +37,75 @@ func connectWebSocket(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	wc := c.App.NewWebConn(ws, *c.App.Session(), c.App.T, "")
-
+	var connID string
+	var seq int
+	var activeQueue chan model.WebSocketMessage
+	var deadQueue []*model.WebSocketEvent
+	var deadQueuePointer int
+	var active bool = true
 	if *c.App.Config().ServiceSettings.EnableReliableWebSockets {
-		connID := r.URL.Query().Get(connectionIDParam)
-		if connID == "" {
+		connID = r.URL.Query().Get(connectionIDParam)
+		if connID == "" || c.App.Session().UserId == "" {
 			// If not present, we assume client is not capable yet, or it's a fresh connection.
 			// We just create a new ID.
 			connID = model.NewId()
+			// In case of fresh connection id, sequence number is already zero.
 		} else {
 			if !model.IsValidId(connID) {
-				mlog.Error("Invalid connection ID", mlog.String("id", connID))
-				wc.WebSocket.Close()
+				mlog.Warn("Invalid connection ID", mlog.String("id", connID))
+				ws.Close()
 				return
 			}
-			// If present, we check if it's present in the connection manager.
-			// TODO: the connection manager internally should forward the request
+
+			// TODO: the method should internally forward the request
 			// to the cluster if it does not have it.
-			//
-			// If the connection is not present, then we assume either timeout,
-			// or server restart. In that case, we set a new one.
-			//
-			// Now we get the sequence number
-			seqVal := r.URL.Query().Get(sequenceNumberParam)
-			if seqVal == "" {
-				// Sequence_number must be sent with connection id.
-				// A client must be either non-compliant or fully compliant.
-				mlog.Error("Sequence number not present in websocket request")
-				wc.WebSocket.Close()
-				return
+			res := c.App.CheckWebConn(c.App.Session().UserId, connID)
+			if res == nil {
+				// If the connection is not present, then we assume either timeout,
+				// or server restart. In that case, we set a new one.
+				connID = model.NewId()
+			} else {
+				// Connection is present, we get the active queue, dead queue
+				activeQueue = res.ActiveQueue
+				deadQueue = res.DeadQueue
+				deadQueuePointer = res.DeadQueuePointer
+				active = false
+				// Now we get the sequence number
+				seqVal := r.URL.Query().Get(sequenceNumberParam)
```

Comment:
Oh yeah .. I thought long about this one. I settled down on this approach after my third iteration :P

The issue with reviving the queue in-place is that the registration code gets unnecessarily complicated. You will have to drain the dead queue _before_ starting to send the active queue. But hub can only send to the active queue. So the best compromise I came up was to initialize the webconn with either fresh data, or reused data, and clear out stale data if any. And then web hub can send as usual.

This way most complexity remains inside webconn, rather than spread all over the place.

I think this looks a bit odd because we are throwing the data all the way from hub to the api layer, and then coming back to app layer again. I didn't think too much about this because I have been looking at this code for way too long :P But I'll clear my mind and probably move some of the initialization step inside webconn. That way the overall logic should look a bit cleaner.